### PR TITLE
[GJ-100] Enable Velox backend on master branch

### PR DESCRIPTION
**Pull request by rui-mo**
This pr:

- ported the code from velox_dev branch to master, and enabled Velox as a backend.
- used ReadRel to replace InputRel.

Add below mvn options to compile Gluten with Velox backend:
`-Dvelox_home=${VELOX_HOME} -Dbuild_velox=ON`

Depends on Arrow PR: https://github.com/oap-project/arrow/pull/94.
Depends on Velox branch: https://github.com/rui-mo/velox/commits/velox_for_gazelle_jni.
Closes https://github.com/oap-project/gluten/issues/100.

### CODE REVIEW

File: jvm/src/main/scala/org/apache/spark/sql/execution/ColumnarCollapseCodegenStages.scala [hunk 1]
```diff
@@ -136,7 +136,14 @@ case class ColumnarCollapseCodegenStages(
...
             codegenStageCounter.incrementAndGet())
       case other =>
-        other.withNewChildren(other.children.map(insertWholeStageTransformer))
+        val maybeWS = other.children.map(insertWholeStageTransformer)
+        // Fake Arrow format will be returned for WS transformer if the next operator
+        // is ArrowColumnarToRowExec.
+        if (other.isInstanceOf[ArrowColumnarToRowExec] &&
+            maybeWS.head.isInstanceOf[WholeStageTransformerExec]) {
+          maybeWS.head.asInstanceOf[WholeStageTransformerExec].setFakeOutput()
+        }
```
Comment (rui-mo) [inline jvm/src/main/scala/org/apache/spark/sql/execution/ColumnarCollapseCodegenStages.scala:145]: Behave according to config.

**Comment by rui-mo** [comment]
Will use ReadRel to replace InputRel.

**Comment by rui-mo** [comment]
@marin-ma please help review this PR. thanks.

**Comment by zzcclp** [comment]
Verified with ClickHouse backend. LGTM.

**Pull request merged**